### PR TITLE
Implement cupy.partition

### LINKED ISSUE #162: Implement cupy.partition
`cupy.partition`, the counterpart of `numpy.partition`, creates a copy of an array with its elements rearranged in such a way that the value of the element in k-th position is in the position it would be in a sorted array. This function is equivalent to `std::nth_element` of C++ STL.

numpy.partition
https://docs.scipy.org/doc/numpy/reference/generated/numpy.partition.html

std::nth_element
http://en.cppreference.com/w/cpp/algorithm/nth_element

Unfortunately, neither of the following parallel algorithm libraries implement this kind of parallel selection algorithm:

- Thrust (https://github.com/thrust/thrust)
- Modern GPU (https://github.com/moderngpu/moderngpu)
- CUB (https://nvlabs.github.io/cub/)

For now, the most likely is to implement radix select based on CUB's radix sort implementation.

### PR DESCRIPTION
**Pull request by takagi**
Merge #229 first.

This PR implements `cupy.partition` to fix #162 .

### CODE REVIEW

**Comment by takagi** [comment]
I remove [WIP].

**Comment by takagi** [comment]
jenkins, test this please.

**Comment by takagi** [comment]
jenkins, test this please.

**Comment by takagi** [comment]
jenkins, test this please.

**Comment by takagi** [comment]
jenkins, test this please.

**Comment by takagi** [comment]
CI passed except unstable failure of a test not on this PR.

File: cupy/core/core.pyx [hunk 1]
```diff
@@ -824,7 +824,44 @@ cdef class ndarray:
         else:
             return cupy.rollaxis(idx_array, -1, axis)
 
-    # TODO(okuta): Implement partition
+    def partition(self, kth, axis=-1):
+        """Partially sorts an array
```
Comment (unnonouno) [inline cupy/core/core.pyx]: Add a period.

File: cupy/core/core.pyx [hunk 2]
```diff
@@ -824,7 +824,44 @@ cdef class ndarray:
...
+
+        Args:
+            kth (int or sequence of ints): Element index to partition by. If
+                supplied with a sequence of k-th it will partition all elements
+                indexed by k-th of them into their sorted position at once.
+            axis (int): Axis along which to sort. Default is -1, which means
+                sort along the last axis.
+
+        .. note::
+           For its implementation reason, `cupy.ndarray.partition` fully sorts
```
Comment (unnonouno) [inline cupy/core/core.pyx]: Use `:func:`

File: cupy/core/core.pyx [hunk 3]
```diff
@@ -824,7 +824,44 @@ cdef class ndarray:
...
+           the given array as `cupy.ndarray.sort` does. It also does not
+           support ``kind`` and ``order`` parameters that ``numpy.partition``
+           supports.
+
+        .. seealso::
+            :func: `cupy.partition` for full documentation,
```
Comment (unnonouno) [inline cupy/core/core.pyx]: Remove a space between `:func:` and `cupy.partition`

File: cupy/core/core.pyx [hunk 4]
```diff
@@ -824,7 +824,44 @@ cdef class ndarray:
...
+            :meth:`numpy.ndarray.partition`
+
+        """
+        ndim = self.ndim
+        if axis < 0:
+            axis += ndim
+        if not (0 <= axis < ndim):
+            raise ValueError('Axis out of range')
+
+        len = self.shape[axis]
```
Comment (unnonouno) [inline cupy/core/core.pyx]: Don't use reserved word. How about `length`?

File: cupy/core/core.pyx [hunk 5]
```diff
@@ -824,7 +824,44 @@ cdef class ndarray:
...
+        kth = (kth,) if isinstance(kth, int) else kth
```
Comment (unnonouno) [inline cupy/core/core.pyx]: How about like this?
```
if isinstance(kth, int):
    kth = kth,
```

File: cupy/sorting/sort.py [hunk 6]
```diff
@@ -127,7 +127,41 @@ def msort(a):
...
+        axis (int or None): Axis along which to sort. Default is -1, which
+            means sort along the last axis. If None is supplied, the array is
+            flattened before sorting.
+
+    Returns:
+        cupy.ndarray: Array of the same type and shape as ``a``.
+
+    .. note::
+       For its implementation reason, `cupy.partition` fully sorts the given
+       array as `cupy.sort` does. It also does not support ``kind`` and
```
Comment (unnonouno) [inline cupy/sorting/sort.py]: use `:func:`

**Comment by takagi** [comment]
I fixed them!

File: cupy/core/core.pyx [hunk 7]
```diff
@@ -824,7 +824,45 @@ cdef class ndarray:
...
+        Args:
+            kth (int or sequence of ints): Element index to partition by. If
+                supplied with a sequence of k-th it will partition all elements
+                indexed by k-th of them into their sorted position at once.
+            axis (int): Axis along which to sort. Default is -1, which means
+                sort along the last axis.
+
+        .. note::
+           For its implementation reason, :func:`cupy.ndarray.partition` fully
+           sorts the given array as `cupy.ndarray.sort` does. It also does not
```
Comment (unnonouno) [inline cupy/core/core.pyx]: `:meth:`

File: cupy/core/core.pyx [hunk 8]
```diff
@@ -824,7 +824,45 @@ cdef class ndarray:
...
+           support ``kind`` and ``order`` parameters that ``numpy.partition``
```
Comment (unnonouno) [inline cupy/core/core.pyx]: `:func:`

File: cupy/sorting/sort.py [hunk 9]
```diff
@@ -127,7 +127,41 @@ def msort(a):
...
+            indexed by k-th of them into their sorted position at once.
+        axis (int or None): Axis along which to sort. Default is -1, which
+            means sort along the last axis. If None is supplied, the array is
+            flattened before sorting.
+
+    Returns:
+        cupy.ndarray: Array of the same type and shape as ``a``.
+
+    .. note::
+       For its implementation reason, `cupy.partition` fully sorts the given
```
Comment (unnonouno) [inline cupy/sorting/sort.py]: `:func:`

File: cupy/sorting/sort.py [hunk 10]
```diff
@@ -127,7 +127,41 @@ def msort(a):
...
+       array as :func:`cupy.sort` does. It also does not support ``kind`` and
+       ``order`` parameters that ``numpy.partition`` supports.
```
Comment (unnonouno) [inline cupy/sorting/sort.py]: `:func:`

**Comment by takagi** [comment]
Fixed!

**Comment by unnonouno** [comment]
LGTM!

**Pull request merged**